content/schema-evolution-field-operations/index.md

```diff
@@ -0,0 +1,158 @@
+---
+rfc:
+start_date: 2017-11-17
+status: draft
+---
+
+# Schema Evolution: Add field operation
+
+## Summary
+
+This RFC proposes a way to add fields to a Register, and defines the meaning
+of this operation. Thus, it doesn’t attempt to cover other field operations
+such as field subtraction or field transformation (e.g.  change the datatype
+of a field).
+
+Note that you can find the [glossary](#glossary) of concepts used throughout
+this RFC at the end of it.
+
+Note: The pseudocode in this RFC resembles [Elm](http://elm-lang.org) but it
+is not intended to be valid.
+
+
+## Motivation
+
+Currently, the Registers Data Model has no way to evolve the data structure
+described in the schema and used in data items. This means a Register has to
+be designed upfront with the right fields forever given it is an append-only
+log. We need a way to be able to evolve the schema definition and a way to
+reflect such evolution on existing and new data.
+
+
+## Explanation
+
+
+### Field addition
+
+The addition of a field, [`Schema.add`](#schema-add-function), adds a
+[field](#field) to the [schema](#schema).
+
+
+##### Example: Add a field to the schema
+
+```elm
+s_0 = Schema [("x", Number)]
+s_1 = Schema.add s_0 ("y", Number) -- s_1 == Schema [("x", Number), ("y", Number)]
+```
+
+
+### Item addition
+
+The item addition operation, [`Item.add`](#item-add-function), adds the item
+if [`Item.validate`](#item-validate-function) succeeds. This RFC doesn't
+change its behaviour, the only thing to keep in mind is that the validation
+function checks against the current schema.
+
+The addition of an item steps are:
+
+1. Validate `a` against the current schema `s`: `Item.validate s a`.
+    * If the item is invalid, reject the operation.
+2. Canonicalise the item `a_p`: `a_c = Item.canonicalise a_p`.
+3. Add item `a_c` to the Register.
+
+
+### REST API Implications
+
+The REST API has no changes given that it is a read-only API.
```

Comment:
Can it cope with old items that don't have the new field?

Reply:
Yes, the assumption in a Register (previous to this RFC) is that a consumer has to be able to handle any response serialised in the requested format (e.g. json) and expect to receive part of the known fields. Also it has to ignore unknown fields. That is:

Given a consumer that knows about a schema such as `Schema [("x", Number), ("y", Number), ("z", String)]` and assuming `x` is the primary field, a consumer has to be able to expect and handle:

#### All known fields present

Given `{"x": 1, "y": 2, "z": "foo"}`, keep `{"x": 1, "y": 2, "z": "foo"}`.

#### Some known fields present

Given `{"x": 1, "y": 2}`, keep `{"x": 1, "y": 2}` and not fail because the missing `z`.

#### Some unknown fields present

Given `{"x": 1, "y": 2, "z": "foo", "n": 42}`, keep `{"x": 1, "y": 2, "z": "foo"}` and not fail because the unknown `n`.

Arguably in the last case, the consumer may want to check if there is a new schema and act accordingly but it depends on the implementation.

Reply:
When the serialisation has to keep columnar consistency (e.g. CSV), then items previous to the new field would have a missing value there. The meaning of this missing value is the `null` as "unknown".

This RFC doesn't add any mechanism for defining default values other than `null` because it is a complex topic on its own that deserve full attention to be sure we are choosing the right tradeoffs.

Reply:
For clarification on the last point of **Some unknown fields present**, are we talking about the consumer reading from the registers API here or adding to the register (I'm also the former from the remainder of the description, but a little confused here).

Given `{"x": 1, "y": 2, "z": "foo", "n": 42}`, the current Java implementation of the app would disallow the Item because of an unknown field for the register, which here is `n`. Conversely, if this Item were already part of the register and the consumer was reading from the register when field `n` was not part of the register, then this would be a case of field removal rather than addition.

The current Java implementation may just be that, and it may be perfectly acceptable for a register to accept an incoming Item with an unknown field, but just want to double-check this here. Apologies if I've missed the point!

Reply:
@karlbaker02 the points I was talking about are from the point of view of the API consumer (which is read-only at all effects). So, from the point of view of the current Java implementation (API provider) `n` would be known and hence the item valid. I'm not following the "field removal" part I'm afraid.